tests/testhooks.nim

```diff
@@ -77,12 +77,26 @@ proc serverHook(): Hook =
     verify: serverVerify
   )
 
+proc serverVerifyWithCode(ctx: Hook, headers: HttpTable):
+     Result[HttpCode, string] {.gcsafe, raises: [Defect].} =
+  var p = TokenHook(ctx)
+  if headers.getString("auth-token") == "good-token":
+    p.status = 101
+    ok(Http101)
+  else:
+    ok(Http401) # Unauthorized
```

Comment:
To illustrate the case, I've made an example returning HttpCode from hook `verify` result. but It feels weird.
```nim
  Hook* = ref object of RootObj
    append*: proc(ctx: Hook,
                  headers: var HttpTable): Result[void, string]
                  {.gcsafe, raises: [Defect].}
    verify*: proc(ctx: Hook,
                  headers: HttpTable): Result[void, string]
                  headers: HttpTable): Result[HttpCode, string]     # weird
                  {.gcsafe, raises: [Defect].}
```

Second approach also not very intuitive:
```Nim
Hook* = ref object of RootObj
    append*: proc(ctx: Hook,
                  headers: var HttpTable): Result[void, string]
                  {.gcsafe, raises: [Defect].}
    verify*: proc(ctx: Hook,
                  headers: HttpTable): Result[void, string]
                  headers: HttpTable): Result[void, (HttpCode, string)]  # also not very good
                  {.gcsafe, raises: [Defect].}
```

Third approach is less descriptive:
```Nim
Hook* = ref object of RootObj
    append*: proc(ctx: Hook,
                  headers: var HttpTable): Result[void, string]
                  {.gcsafe, raises: [Defect].}
    verify*: proc(ctx: Hook,
                  headers: HttpTable): Result[void, string]
                  headers: HttpTable): Result[void, HttpCode]  # limiting the error message to HttpCode
                  {.gcsafe, raises: [Defect].}
```

Reply:
I think that there is no need to communicate anything back to the caller of the hook because we aren't doing anything with it at all... 

The hooks are application specific and executed within the application context they where declared in. Anything application specific related to the failed verification should happen in the hook itself. The only reasonable return value from the hook is a straight forward `bool` that indicates whether connection establishing should proceed or halt.

Reply:
For logging purposes, we might return a `string`, just as in one of your examples.

Reply:
okay. I think I get the idea. Will clean it up and add some docs.

Reply:
I think this `Result[void, string]` would be a reasonable return value, given that we want a string for logging purposes.